source/simulator/lateral_averaging.cc

```diff
@@ -312,22 +360,54 @@ namespace aspect
                                              std::vector<double>(n_slices,0.0));
     std::vector<double> volume(n_slices,0.0);
 
-    // this yields 10^dim quadrature points evenly distributed in the interior of the cell.
-    // We avoid points on the faces, as they would be counted more than once.
-    const QIterated<dim> quadrature_formula (QMidpoint<1>(),
-                                             10);
-    const unsigned int n_q_points = quadrature_formula.size();
+    // We would like to use a quadrature formula that is appropriately accurate laterally,
+    // but has a high resolution in depth (to populate all depth slices for the averages).
+    // For that we need to know the depth direction in the unit cell coordinate system, which
+    // is only unique (= the same for all cells) in some geometries. In these geometries we
+    // can optimize the quadrature, otherwise we need to use a high-resolution quadrature in
+    // all directions, which is more expensive.
+    const bool geometry_model_has_unique_depth_direction = Plugins::plugin_type_matches<GeometryModel::Box<dim> >(this->get_geometry_model()) ||
+                                                           Plugins::plugin_type_matches<GeometryModel::Chunk<dim> >(this->get_geometry_model()) ||
+                                                           Plugins::plugin_type_matches<GeometryModel::EllipsoidalChunk<dim> >(this->get_geometry_model()) ||
+                                                           Plugins::plugin_type_matches<GeometryModel::SphericalShell<dim> >(this->get_geometry_model()) ||
+                                                           Plugins::plugin_type_matches<GeometryModel::TwoMergedBoxes<dim> >(this->get_geometry_model());
+
+    unsigned int depth_direction = numbers::invalid_unsigned_int;
+
+    // The Chunk geometry model has depth as first dimension (radius, lon, lat),
+    // all others with unique direction have it last.
+    if (geometry_model_has_unique_depth_direction)
+      {
+        depth_direction = (Plugins::plugin_type_matches<GeometryModel::Chunk<dim> >(this->get_geometry_model()))
+                          ?
+                          1
+                          :
+                          dim;
+      }
+
+    const unsigned int max_fe_degree = std::max(this->introspection().polynomial_degree.velocities,
+                                                std::max(this->introspection().polynomial_degree.temperature,
+                                                         this->introspection().polynomial_degree.compositional_fields));
+
+    std::unique_ptr<Quadrature<dim> > quadrature_formula;
+    if (geometry_model_has_unique_depth_direction)
+      quadrature_formula = std::make_unique<Quadrature<dim> >(internal::get_quadrature_formula<dim>(max_fe_degree+1,
```

Comment:
can you please leave a comment for the ``max_fe_degree+1`` logic? I assume that the +1 is also quite conservative, but maybe we should leave it here (curved geometry).